datastore/providers/opensearch_datastore.py

```diff
@@ -0,0 +1,422 @@
+import os
+import json
+
+
+import boto3
+from opensearchpy import AsyncOpenSearch, AWSV4SignerAsyncAuth, AsyncHttpConnection
+from requests_kerberos import HTTPKerberosAuth, OPTIONAL
+
+from typing import Any, Dict, List, Optional
+from datastore.datastore import DataStore
+from models.models import (
+    DocumentChunk,
+    DocumentChunkWithScore,
+    DocumentMetadataFilter,
+    QueryResult,
+    QueryWithEmbedding, DocumentChunkMetadata,
+    Source
+)
+from services.date import validate_date_str
+
+# Read environment variables for OpenSearch
+OPENSEARCH_HOST = os.environ.get("OPENSEARCH_HOST", "localhost")
+OPENSEARCH_PORT = int(os.environ.get("OPENSEARCH_PORT", 9200))
+OPENSEARCH_USERNAME = os.environ.get("OPENSEARCH_USERNAME")
+OPENSEARCH_PASSWORD = os.environ.get("OPENSEARCH_PASSWORD")
+OPENSEARCH_AUTH_TYPE = os.environ.get("OPENSEARCH_AUTH_TYPE", "no-auth")
+OPENSEARCH_INDEX_NAME = os.environ.get("OPENSEARCH_INDEX_NAME", "open-ai-index")
+OPENSEARCH_KNN_FIELD_NAME = os.environ.get("OPENSEARCH_KNN_FIELD_NAME", "embedding")
+OPENSEARCH_INDEX_PRIMARIES = int(os.environ.get("OPENSEARCH_INDEX_PRIMARIES", 1))
+OPENSEARCH_INDEX_REPLICAS = int(os.environ.get("OPENSEARCH_INDEX_REPLICAS", 0))
+OPENSEARCH_KNN_ENGINE = os.environ.get("OPENSEARCH_KNN_ENGINE", "nmslib")
+OPENSEARCH_KNN_VECTOR_DISTANCE = os.environ.get("OPENSEARCH_KNN_VECTOR_DISTANCE", "l2")
+OPENSEARCH_EMBEDDING_IN_RESULT = bool(os.environ.get("OPENSEARCH_EMBEDDING_IN_RESULT", False))
+CA_CERTS_FULL_PATH = os.environ.get("CA_CERTS_FULL_PATH")
+AWS_REGION = os.environ.get("AWS_REGION", "us-east-2")
```

Comment:
Why us-east-2?

Reply:
the region is required for awsSigv4. I just added a random region of AWS.